File: c++/leetcode/1. Two Sum.cpp
#include <vector>
using std::vector;
#include <unordered_map>
using std::unordered_map;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> hash; // {value: index}
        vector<int> result = {-1, -1};
        for (int i = 0; i < nums.size(); i++) {
            int numberToFind = target - nums.at(i);
            if (hash.find(numberToFind) != hash.end()) {
                result = {hash[numberToFind], i};
                return result;
            }
            hash[nums.at(i)] = i;
        }
        return result;
    }
};

File: ts/leetcode/1. Two Sum.ts
// brute force: O(n ^ 2)
function twoSum1(nums: number[], target: number): number[] {
    if (!nums) {
        return [-1, -1];
    }

    for (let i = 0; i < nums.length; i ++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) {
                return [i, j];
            }
        }
    }

    return [-1, -1];
};

// hash table:
function twoSum(nums: number[], target: number): number[] {
    const hashtable: {[index: number]: number} = {};
    for (let i = 0; i < nums.length; i++) {
        const diff = target - nums[i];
        if (hashtable[diff] !== undefined) return [hashtable[diff], i];
        hashtable[nums[i]] = i;
    }
    return [-1, -1];
}